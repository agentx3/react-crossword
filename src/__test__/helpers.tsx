// import { jest } from '@jest/globals';
import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// dummy to make "use default export!" warning go away.
/* istanbul ignore next */
export function NYI() {
  // eslint-disable-next-line no-console
  console.log('NYI!');
}

export function setup(ui: React.ReactElement) {
  return {
    user: userEvent.setup(),
    ...render(ui),
  };
}
