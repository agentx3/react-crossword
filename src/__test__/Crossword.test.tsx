import { jest } from '@jest/globals';
import React from 'react';
// import { createRoot } from 'react-dom/client';
import { act, render, cleanup, fireEvent } from '@testing-library/react';
// import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import '@testing-library/jest-dom/extend-expect';

import { setup } from './helpers';

import Crossword, { CrosswordImperative } from '../Crossword';

afterEach(cleanup);

const emptyData = {
  across: {},
  down: {},
};

const simpleData = {
  across: {
    1: {
      clue: 'one plus one',
      answer: 'TWO',
      row: 0,
      col: 0,
    },
  },
  down: {
    2: {
      clue: 'three minus two',
      answer: 'ONE',
      row: 0,
      col: 2,
    },
  },
};

const defaultProps = {
  data: emptyData,
  useStorage: false,
};

it('renders without crashing', () => {
  // const div = document.createElement('div');
  // const root = createRoot(div);
  // root.render(<Crossword {...defaultProps} />);
  // root.unmount();
  render(<Crossword {...defaultProps} />);
});

it('renders Crossword component correctly', () => {
  const { container, getByText } = render(<Crossword {...defaultProps} />);
  expect(container.firstChild).toHaveClass('crossword');
  expect(getByText('ACROSS')).toHaveTextContent('ACROSS');
  expect(getByText('DOWN')).toHaveTextContent('DOWN');
});

it('renders Crossword component correctly when using storage', () => {
  const { container, getByText } = render(
    <Crossword {...defaultProps} useStorage />
  );
  expect(container.firstChild).toHaveClass('crossword');
  expect(getByText('ACROSS')).toHaveTextContent('ACROSS');
  expect(getByText('DOWN')).toHaveTextContent('DOWN');
});

it('matches snapshot', () => {
  const tree = renderer.create(<Crossword {...defaultProps} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('creates new gridData when the data changes', () => {
  const clueMatch = /one plus one/;
  const { queryByText, rerender } = render(<Crossword {...defaultProps} />);
  expect(queryByText(clueMatch)).toBeNull();

  rerender(<Crossword {...defaultProps} data={simpleData} />);
  expect(queryByText(clueMatch)).toBeTruthy();
});

describe('imperative commands (forwarded to CrosswordProvider)', () => {
  it('sets focus when requested', () => {
    const ref = React.createRef<CrosswordImperative>();
    const { getByLabelText, container } = render(
      <Crossword {...defaultProps} data={simpleData} ref={ref} />
    );

    const doc = container.ownerDocument;

    // no focus yet?
    const input = getByLabelText('crossword-input');
    expect(doc.activeElement).not.toBe(input);

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.focus();
    });

    expect(doc.activeElement).toBe(input);
  });

  it('resets data when requested', async () => {
    const ref = React.createRef<CrosswordImperative>();
    const { getByLabelText, queryByText, user } = setup(
      <Crossword {...defaultProps} data={simpleData} ref={ref} />
    );

    await user.click(getByLabelText('clue-1-across'));

    const input = getByLabelText('crossword-input');

    fireEvent.keyDown(input, { key: 'X' });
    let textEl = queryByText('X');
    expect(textEl).toBeTruthy();

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.reset();
    });

    textEl = queryByText('X');
    expect(textEl).toBeFalsy();
  });

  it('fills answers when requested', () => {
    const ref = React.createRef<CrosswordImperative>();
    const { queryByText } = render(
      <Crossword {...defaultProps} data={simpleData} ref={ref} />
    );

    let textEl = queryByText('T');
    expect(textEl).toBeFalsy();

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.fillAllAnswers();
    });

    textEl = queryByText('T');
    expect(textEl).toBeTruthy();
  });

  it('calls onLoadedCorrect after filling answers', () => {
    const onLoadedCorrect = jest.fn();
    const ref = React.createRef<CrosswordImperative>();
    render(
      <Crossword
        {...defaultProps}
        data={simpleData}
        onLoadedCorrect={onLoadedCorrect}
        ref={ref}
      />
    );

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.fillAllAnswers();
    });

    expect(onLoadedCorrect).toBeCalledWith([
      ['across', '1', 'TWO'],
      ['down', '2', 'ONE'],
    ]);
  });

  it('calls onCrosswordCorrect after filling answers', () => {
    const onCrosswordCorrect = jest.fn();
    const ref = React.createRef<CrosswordImperative>();
    render(
      <Crossword
        {...defaultProps}
        data={simpleData}
        onCrosswordCorrect={onCrosswordCorrect}
        ref={ref}
      />
    );

    onCrosswordCorrect.mockClear();
    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.fillAllAnswers();
    });

    expect(onCrosswordCorrect).toBeCalledTimes(1);
    expect(onCrosswordCorrect).toBeCalledWith(true);
  });

  it('returns whether the crossword is correct', () => {
    const ref = React.createRef<CrosswordImperative>();
    render(<Crossword {...defaultProps} data={simpleData} ref={ref} />);
    expect(ref.current).toBeTruthy();
    let isCorrect: boolean | undefined = true;
    act(() => {
      isCorrect = ref.current?.isCrosswordCorrect();
    });

    expect(isCorrect).toBeFalsy();
  });

  it('setGuess() can set a guess', () => {
    const ref = React.createRef<CrosswordImperative>();
    const { queryByText } = render(
      <Crossword {...defaultProps} data={simpleData} ref={ref} />
    );

    let textEl = queryByText('T');
    expect(textEl).toBeFalsy();

    expect(ref.current).toBeTruthy();
    act(() => {
      ref.current?.setGuess(0, 0, 'T');
    });

    textEl = queryByText('T');
    expect(textEl).toBeTruthy();
  });
});
