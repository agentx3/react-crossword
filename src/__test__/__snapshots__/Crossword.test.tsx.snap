// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
[
  <div
    className="sc-bcXHqe gsMPLZ crossword grid"
  >
    <div
      style={
        {
          "margin": 0,
          "padding": 0,
          "position": "relative",
        }
      }
    >
      <svg
        viewBox="0 0 10 10"
      >
        <rect
          fill="rgb(0,0,0)"
          height={10}
          width={10}
          x={0}
          y={0}
        />
      </svg>
      <input
        aria-label="crossword-input"
        autoComplete="off"
        autoCorrect="off"
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        spellCheck="false"
        style={
          {
            "backgroundColor": "transparent",
            "border": 0,
            "caretColor": "transparent",
            "cursor": "default",
            "fontSize": "40.949999999999996px",
            "height": "calc(100% - 4px)",
            "left": "calc(0% + 2px)",
            "margin": 0,
            "padding": 0,
            "position": "absolute",
            "textAlign": "center",
            "textAnchor": "middle",
            "top": "calc(0% + 2px)",
            "width": "calc(100% - 4px)",
          }
        }
        type="text"
        value=""
      />
    </div>
  </div>,
  <div
    className="sc-dkrFOg blVNBR clues"
  >
    <div
      className="direction"
    >
      <h3
        className="header"
      >
        ACROSS
      </h3>
    </div>
    <div
      className="direction"
    >
      <h3
        className="header"
      >
        DOWN
      </h3>
    </div>
  </div>,
]
`;
